oauthproxy.go

```diff
@@ -796,6 +807,113 @@ func (p *OAuthProxy) OAuthCallback(rw http.ResponseWriter, req *http.Request) {
 	}
 }
 
+// AuthorizeSession will check the session state and ensure that it meets
+// all of the authorization criteria (email domains and/or claims criteria).
+// If it hasn't
+func (p *OAuthProxy) AuthorizeSession(s *sessionsapi.SessionState, saveSession *bool) (authorized bool, reason string) {
+
+	if !p.Validator(s.Email) {
+		return false, "failed email validation"
+	}
+
+	if !p.provider.ValidateGroup(s.Email) {
+		return false, "failed group validation"
+	}
+
+	return p.ValidateAuthorizedClaims(s, saveSession)
+}
+
+const (
+	authZPass byte = 'P'
+	authZFail byte = 'F'
+)
+
+func (p *OAuthProxy) createAuthzDigest(s *sessionsapi.SessionState, authZresult byte) ([]byte, error) {
+
+	// The digest is keyed to the following:
+	//   - The hash of the proxy's configured authorization rules
+	//   - Pass/Fail the authorization test
+	//   - UserId
+
+	mac := hmac.New(sha256.New, p.claimsAuthorizer.RulesHash())
+	mac.Write([]byte{authZresult})
+	mac.Write([]byte(s.User))
+	return mac.Sum(nil), nil
+}
+
+// ValidateAuthorizedClaims will extract any claims from the idtoken and check them
+// against the rules configured to allow acceptance.
+//  - If no assertions were specified, it will trivially accept any (or no) claims.
+//  - Otherwise, the first assertion that evaluates to a ["truthy"](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) result will return true.
+//  - If no assertion matches, the claims are not deemed acceptable and false is returned.
+func (p *OAuthProxy) ValidateAuthorizedClaims(s *sessionsapi.SessionState, saveSession *bool) (authorized bool, reason string) {
+
+	if p.claimsAuthorizer.IsEmpty() {
+		return true, "*"
+	}
+
+	if s.Authz != "" {
+		// We have a cached claims assertions result so we only need to re-validate
+		// the request if the digest doesn't match.
+
+		// If any of those things changes, the cache is considered invalid and the full
+		// authorization check must be made again.
+
+		if providedMAC, err := base64.StdEncoding.DecodeString(s.Authz[1:]); err == nil {
+			if expectedMAC, err := p.createAuthzDigest(s, s.Authz[0]); err == nil {
+				if hmac.Equal(providedMAC, expectedMAC) {
+					// Note: The cached result is indeed valid, now actually see if we were authorized. :)
+					return s.Authz[0] == authZPass, "(cached)"
+				}
+			}
+
+			// MAC fail (probably) means something changed since we last saw this user (or
+			// the session was just refreshed), we need to re-verify the authorization rules.
+			logger.Printf("Clearing session authz cache and revalidating claims")
+		} else {
+			logger.Printf("Clearing session authz cache because of decode error: %v", err)
+		}
```

Comment:
This to me feels like it could be made tidier in a separate function that checks the cache, WDYT?

Reply:
That'd just force you to look elsewhere for logic that is intrinsic to this function. Nobody else should be validating with it.

That said, I am thinking I will likely provide a way in the future to extract a portion of the claims to embed in the session so they can be sent to the client and/or upstream if desired. In which case we'd want to include them in this digesting mechanism and so it'll probably wind up changing anyway. 😛